### PR TITLE
[Form] Add "choice_translation_parameters" option

### LINKED ISSUE #36845: ChoiceType Field should have a choice_translation_parameters option.
**Description**  

There is a `label_translation_parameters` option (cf https://symfony.com/doc/current/reference/forms/types/choice.html#label-translation-parameters). In the same way, there is a `attr_translation_parameters` and `help_translation_parameters`.

IMHO, a `choice_translation_parameters` option could be great.

**Example**  
```
# translations/messages.en.yaml
form.order.ok: 'I want to confirm the order made to %company%'
form.order.ko: 'I want to cancel the order made to %company%'
```

```
$builder->add('order', null, [
    'choices' => [
        'form.order.ok' => true,
        'form.order.ko' => false,
    ],
    'choice_translation_parameters' => [
        '%company%' => 'ACME Inc.',
    ],
]);
```

### PR DESCRIPTION
**Pull request by VincentLanglet**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | Fix #36845 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/13677

Original PR: https://github.com/symfony/symfony/pull/36851

### CODE REVIEW

**Comment by VincentLanglet** [comment]
As explained here: https://github.com/symfony/symfony/pull/36851#issuecomment-703284784

I don't know how to avoid the deprecation in the tests (that's why they are failing).
That's the last thing before being RTM.

If you can help me @fabpot @xabbuh @nicolas-grekas.

**Comment by VincentLanglet** [comment]
As explained here: https://github.com/symfony/symfony/pull/36851#issuecomment-703284784

I don't know how to avoid the deprecation in the tests (that's why they are failing).
That's the last thing before being RTM.

Could you help me @fabpot @xabbuh @nicolas-grekas ? Thanks.

**Comment by VincentLanglet** [comment]
@fabpot @xabbuh @nicolas-grekas This PR is ready to merge except the deprecation in the tests.
But I don't know how to get rid of it... Could you give me some help ?

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php [hunk 1]
```diff
@@ -69,8 +69,9 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, callable $va
     /**
      * {@inheritdoc}
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, callable $index = null, callable $groupBy = null, $attr = null)
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, callable $index = null, callable $groupBy = null, $attr = null/*, $labelTranslationParameters = []*/)
```
Comment (yceruto) [inline src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php:74]: I guess our `DebugClassLoader` is not smart enough yet to detect that the new parameter doc is being added in the parent docblock. So you'll need to add it here to get rid of the current deprecation notice you're facing.

Comment (VincentLanglet) [reply src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php:74]: Thanks ! I would have never find this !

**Comment by VincentLanglet** [comment]
I rebased, it should be RTM

File: src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php [hunk 2]
```diff
@@ -174,14 +174,16 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, $value = nul
...
+     * @param callable|false|Cache\ChoiceLabel|null            $label                      The option or static option generating the choice labels
+     * @param callable|Cache\ChoiceFieldName|null              $index                      The option or static option generating the view indices
+     * @param callable|Cache\GroupBy|null                      $groupBy                    The option or static option generating the group names
+     * @param array|callable|Cache\ChoiceAttr|null             $attr                       The option or static option generating the HTML attributes
+     * @param array|callable|Cache\ChoiceTranslationParameters $labelTranslationParameters The parameters used to translate the choice labels
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null)
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null/*, $labelTranslationParameters = []*/)
     {
+        $labelTranslationParameters = \func_get_args()[6] ?? [];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: better use func_num_args

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php [hunk 3]
```diff
@@ -68,9 +68,12 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, callable $va
 
     /**
      * {@inheritdoc}
+     *
+     * @param array|callable $labelTranslationParameters The parameters used to translate the choice labels
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, callable $index = null, callable $groupBy = null, $attr = null)
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, callable $index = null, callable $groupBy = null, $attr = null/*, $labelTranslationParameters = []*/)
     {
+        $labelTranslationParameters = \func_get_args()[6] ?? [];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: better use func_num_args

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php:74]: it's not that it's not smart enough - it's that being explicit is needed here, to prove that you cared about adding the argument, event commented

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php [hunk 4]
```diff
@@ -186,7 +191,12 @@ private static function addChoiceView($choice, string $value, $label, array $key
             $label,
             // The attributes may be a callable or a mapping from choice indices
             // to nested arrays
-            \is_callable($attr) ? $attr($choice, $key, $value) : (isset($attr[$key]) ? $attr[$key] : [])
+            \is_callable($attr) ? $attr($choice, $key, $value) : (isset($attr[$key]) ? $attr[$key] : []),
+            // The label translation parameters may be a callable or a mapping from choice indices
+            // to nested arrays
+            \is_callable($labelTranslationParameters)
+                ? $labelTranslationParameters($choice, $key, $value)
+                : (isset($labelTranslationParameters[$key]) ? $labelTranslationParameters[$key] : [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: should be on one line + `($labelTranslationParameters[$key] ?? [])`

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php [hunk 5]
```diff
@@ -186,7 +191,12 @@ private static function addChoiceView($choice, string $value, $label, array $key
             $label,
             // The attributes may be a callable or a mapping from choice indices
             // to nested arrays
-            \is_callable($attr) ? $attr($choice, $key, $value) : (isset($attr[$key]) ? $attr[$key] : [])
+            \is_callable($attr) ? $attr($choice, $key, $value) : (isset($attr[$key]) ? $attr[$key] : []),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php]: `$attr[$key] ?? []` while at it

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 6]
```diff
@@ -145,16 +145,18 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, $value = nul
...
+     * @param callable|string|PropertyPath|null       $groupBy                    The callable or path generating the group names
+     * @param array|callable|string|PropertyPath|null $attr                       The callable or path generating the HTML attributes
+     * @param array|callable|string|PropertyPath      $labelTranslationParameters The callable or path generating the parameters used to translate the choice labels
      *
      * @return ChoiceListView The choice list view
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null)
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null/*, $labelTranslationParameters = []*/)
     {
+        $labelTranslationParameters = \func_get_args()[6] ?? [];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: func_num_args

File: src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php [hunk 7]
```diff
@@ -217,6 +219,24 @@ public function createView(ChoiceListInterface $list, $preferredChoices = null,
             };
         }
 
-        return $this->decoratedFactory->createView($list, $preferredChoices, $label, $index, $groupBy, $attr);
+        if (\is_string($labelTranslationParameters)) {
+            $labelTranslationParameters = new PropertyPath($labelTranslationParameters);
+        }
+
+        if ($labelTranslationParameters instanceof PropertyPath) {
+            $labelTranslationParameters = function ($choice) use ($accessor, $labelTranslationParameters) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php]: should be static I suppose

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 8]
```diff
@@ -12,6 +12,7 @@ CHANGELOG
  * Deprecated passing an array as the first argument of the `CheckboxListMapper::mapFormsToData()` method, pass `\Traversable` instead.
  * Deprecated passing an array as the second argument of the `RadioListMapper::mapDataToForms()` method, pass `\Traversable` instead.
  * Deprecated passing an array as the first argument of the `RadioListMapper::mapFormsToData()` method, pass `\Traversable` instead.
+ * Added a `choice_translation_parameters` option to `ChoiceType`
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/CHANGELOG.md:15]: should it be named `label_translation_parameters`?

Comment (VincentLanglet) [reply src/Symfony/Component/Form/CHANGELOG.md:15]: `label_translation_parameters` already exists and is used for the `label`.

This option is the same but for `choices`.

File: src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php:186]: You mean 
```
        if (\func_num_args() >= 6) {
            $labelTranslationParameters = \func_get_args()[6];
        } else {
            $labelTranslationParameters = [];
        }
```
?

Comment (nicolas-grekas) [reply src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: Yes. More exactly (as found elsewhere in the codebase):
`$labelTranslationParameters = 5 < \func_num_args() ? func_get_arg(5) : [];`

Comment (nicolas-grekas) [reply src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php]: (with the correct index/offset :) )

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php (hunk 1 above)
Comment (yceruto) [reply src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php:74]: It makes sense, thank you for the explanation.

**Review by yceruto — APPROVED** [review]
LGTM!

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @VincentLanglet.

**Pull request merged**